File: packages/cloud115/src/cloud115/rate_limit.py
"""QPS/QPM 限流器，状态持久化到 SQLite。"""
from __future__ import annotations

import threading
import time

from cloud115.log import get_logger


class RateLimiter:
    """限流器。状态通过 FileCache 的 rate_limit 表持久化。

    cache=None 时禁用持久化（纯进程内限流）。
    """

    def __init__(self, name, qps=0.5, qpm=20, cache=None):
        self._name = name
        self._qps = qps
        self._qpm = qpm
        self._cache = cache
        self._lock = threading.Lock()
        self.request_count = 0

    def acquire(self):
        with self._lock:
            self._wait_for_slot()
            self.request_count += 1

    def _wait_for_slot(self):
        if self._cache is None:
            return

        min_interval = 1.0 / self._qps

        while True:
            now = time.time()

            # Atomic check-and-update: single SQL, no TOCTOU race
            if self._cache.try_acquire_slot(self._name, now, min_interval, self._qpm):
                return  # Got a slot

            # Failed — figure out why and wait appropriately
            state = self._cache.get_rate_limit(self._name)

            # Cooldown active?
            if now < state.get("cooldown_until", 0):
                remaining = int(state["cooldown_until"] - now)
                mins, secs = divmod(remaining, 60)
                raise RuntimeError(
                    "Rate limit cooldown: %dm%02ds remaining" % (mins, secs)
                )

            # QPM exceeded?
            minute_start = state.get("minute_start", 0)
            minute_count = state.get("minute_count", 0)
            if now - minute_start <= 60 and minute_count >= self._qpm:
                wait = 60 - (now - minute_start)
                get_logger().debug(
                    "THROTTLE   QPM limit qpm=%d count=%d wait=%.1fs",
                    self._qpm, minute_count, max(wait, 0),
                )
                time.sleep(max(wait, 0.1))
                continue

            # QPS too fast
            last_req = state.get("last_request", 0)
            wait = min_interval - (now - last_req)
            if wait > 0:
                get_logger().debug(
                    "THROTTLE   wait %.1fs qps=%.1f last_req=%.1fs_ago",
                    wait, self._qps, now - last_req,
                )
                time.sleep(wait)
            else:
                time.sleep(0.05)  # Brief pause before retry

    def set_cooldown(self, seconds):
        if self._cache:
            until = time.time() + seconds
            self._cache.set_rate_limit(self._name, cooldown_until=until)
            get_logger().warning(
                "COOLDOWN   %ds until=%.0f", int(seconds), until
            )

File: internal/cloud115/ratelimit.go
package cloud115

import (
	"fmt"
	"log/slog"
	"sync"
	"time"
)

// RateLimiter enforces QPS and QPM limits, persisting state to SQLite via Cache.
// When cache is nil the limiter still tracks request counts but does not block.
type RateLimiter struct {
	name  string
	qps   float64
	qpm   int
	cache *Cache
	mu    sync.Mutex
	count int // total successful acquires (for stats / tests)
}

// NewRateLimiter creates a RateLimiter.
// name   – key used in the rate_limit table (e.g. "api", "download").
// qps    – target queries per second (e.g. 0.5 means one request every 2 s).
// qpm    – maximum requests per rolling 60-second window.
// cache  – pass nil to disable persistence (no blocking beyond mutex).
func NewRateLimiter(name string, qps float64, qpm int, cache *Cache) *RateLimiter {
	return &RateLimiter{
		name:  name,
		qps:   qps,
		qpm:   qpm,
		cache: cache,
	}
}

// Acquire blocks until a slot is available, then increments the request count.
// Returns an error only when a cooldown is active.
func (r *RateLimiter) Acquire() error {
	r.mu.Lock()
	defer r.mu.Unlock()

	if err := r.waitForSlot(); err != nil {
		return err
	}
	r.count++
	return nil
}

// SetCooldown persists a cooldown_until timestamp for this limiter.
// After a 429 response the caller sets a cooldown so the next Acquire will fail
// immediately instead of retrying.
func (r *RateLimiter) SetCooldown(seconds float64) {
	if r.cache == nil {
		return
	}
	until := float64(time.Now().Unix()) + seconds
	r.cache.SetRateLimit(r.name, RateLimitState{CooldownUntil: until})
	slog.Warn("COOLDOWN", "limiter", r.name, "seconds", int(seconds), "until", until)
}

// waitForSlot is the inner loop called with the mutex already held.
func (r *RateLimiter) waitForSlot() error {
	if r.cache == nil {
		return nil
	}

	minInterval := 1.0 / r.qps

	for {
		now := float64(time.Now().UnixNano()) / 1e9

		if r.cache.TryAcquireSlot(r.name, now, minInterval, r.qpm) {
			return nil
		}

		// Failed — inspect why.
		state := r.cache.GetRateLimit(r.name)

		// Cooldown active?
		if now < state.CooldownUntil {
			remaining := int(state.CooldownUntil - now)
			mins := remaining / 60
			secs := remaining % 60
			return fmt.Errorf("rate limit cooldown: %dm%02ds remaining", mins, secs)
		}

		// QPM exceeded?
		minuteStart := state.LastRequest // fallback; cache stores it internally
		_ = minuteStart
		// Re-read full state from DB to decide: the cache row tracks minute_start
		// but GetRateLimit only returns CooldownUntil, LastRequest, MinuteCount.
		// Use LastRequest as minute_start proxy; the real minute_start is in TryAcquireSlot logic.
		// We check MinuteCount against qpm.
		if state.MinuteCount >= r.qpm {
			// Wait until the minute window resets. We estimate based on LastRequest.
			wait := 60.0 - (now - state.LastRequest)
			if wait < 0 {
				wait = 0.1
			}
			slog.Debug("THROTTLE QPM",
				"limiter", r.name,
				"qpm", r.qpm,
				"count", state.MinuteCount,
				"wait_s", fmt.Sprintf("%.1f", wait),
			)
			sleep(wait)
			continue
		}

		// QPS too fast?
		wait := minInterval - (now - state.LastRequest)
		if wait > 0 {
			slog.Debug("THROTTLE QPS",
				"limiter", r.name,
				"qps", r.qps,
				"wait_s", fmt.Sprintf("%.1f", wait),
			)
			sleep(wait)
		} else {
			// Edge case: slot not obtained yet but no obvious reason — brief pause.
			sleep(0.05)
		}
	}
}

// sleep sleeps for the given number of seconds (float).
func sleep(seconds float64) {
	time.Sleep(time.Duration(seconds * float64(time.Second)))
}
